12to11.man
.TH 12to11
.SH NAME
12to11 - Wayland to X protocol translator
.SH SYNOPSIS
.B 12to11
[\-\fBclass\fP \fIclass\fP] [\-\fBname\fP \fIname\fP] [\-\fBxrm\fP \fIresourcestring\fP...]
.SH DESCRIPTION
.I 12to11
starts a Wayland compositor on the next available socket;
Wayland programs will then be displayed through the X server.
.SH OPTIONS
Some X Toolkit library-style command-line arguments are also accepted,
despite the protocol translator being implemented without using a
toolkit.  These options are:
.TP 8
.B \-class \fIclass\fP
This option specifies the resource class under which resources are to
be obtained.  When not set, it defaults to the string ``12to11''.  The
resource class and name used by the protocol translator are \fInot\fP
set as the
.B WM_CLASS
property on windows created by the protocol translator; instead, those
windows get the classes and names assigned to them by their individual
Wayland clients.
.TP 8
.B \-name \fIname\fP
This option specifies the instance name under which resources are to
be obtained.  When not set, it defaults to the executable file name.
.TP
.B \-help\fP
This option causes the protocol translator to print a message
describing options it accepts.  The protocol translator will then exit
after printing the message.
.TP
.B \-xrm\fP \fIresourcestring\fP
This option specifies a resource string to be used.  This is
especially useful for setting resources that do not have separate
command line options.
.TP
.B \-printsocket\fP
This option makes the protocol translator print the name of the
Wayland socket to standard output.
.SH RESOURCES
\fI12to11\fP understands some resource names and classes that can be
used to specify various settings that affect its behavior.  Those
resources are listed below:
.TP 8
.B ximFont\fP (class \fBXimFont\fP)
Specifies the font to be used for displaying the preedit or status
areas of an input method.
.IP
In many input methods, the preedit (a.k.a composition or
preconversion) string and/or a status window is displayed on screen
while typing by the input method server.  The client is supposed to
determine the font used by the input method server to display --
however, the Wayland protocol does not allow Wayland clients to
specify that themselves, so the protocol translator has to do that by
itself.
.TP
.B ximStyles\fP (class \fBXimStyles\fP)
Specifies the default input method styles used by the protocol
translator.
.IP
X input methods can support different editing styles, which affect how
preconversion and status text is displayed while typing.  These styles
are named:
.TP 16
.B overTheSpot
In ``over the spot'', the preedit string is displayed in a window
created by the input method at a location specified by the Wayland
client, typically the text insertion point.
.TP
.B offTheSpot
In ``off the spot'', the preedit string is displayed in a window
created by the input method at some location away from the text
insertion point.  Often, this type of window is placed at the bottom
of the preedit window.
.TP
.B rootWindow
In the ``root window'' editing style, the preedit string is displayed
in a popup window created by the input method.
.TP
.B onTheSpot
In the ``on the spot'' editing style, the preedit string is displayed
inside the text itself by the Wayland client.
.IP "" 8
Not all input methods support all editing styles.  In the real world,
most only support one or two of the styles listed above.  The protocol
translator will search for a style supported by the input method when
initializing input method support.  The \fBximStyles\fP resource is
used to control the order in which the protocol translator searches
for input styles, and should be a comma separated list of input names,
which are searched in left-to-right order.  For example,
.IP
.in +4
.EX
\fBoverTheSpot,rootWindow\fP
.EE
.in
.IP
will result in the protocol translator searching for the ``over the
spot'' input style, and if that is not present, the ``root window''
style.  Whitespace must not be present inside the comma-separated
list.  When \fBximStyles\fP is not specified, it defaults to:
.IP
.in +4
.EX
\fBoverTheSpot,offTheSpot,rootWindow,onTheSpot\fP
.EE
.in
.TP
.B renderer\fP (class \fBRenderer\fP)
Specifies the rendering backend the protocol translator uses to
composite the contents of Wayland surfaces onto X windows.  This can
either be \fBpicture\fP (the XRender based compositor) or \fBegl\fP
(the OpenGL ES 2.0 based compositor).
.TP
.B wmProtocols\fP (class \fBWmProtocols\fP)
Comma-separated list of window manager protocols, similar to
\fBinputStyles\fP, that the protocol translator should enable or
disable.  The current (and default) list of supported protocols are:
\fBnetWmPing\fP.
.TP
.B idleInhibitCommand\fP (class \fBIdleInhibitCommand\fP)
.TP
.B idleIntervalCommand\fP (class \fBIdleInhibitCommand\fP)
.TP
.B idleDeinhibitCommand\fP (class \fBIdleDeinhibitCommand\fP)
These are commands that run when a focused Wayland client asks to
inhibit the screen saver or screen locker, usually because a video is
being played and does not require user interaction.
\fBidleInhibitCommand\fP specifies a command that is run when one such
client becomes focused or starts inhibiting the screen saver;
\fBidleIntervalCommand\fP is a command that is run every several
seconds (by default 60) while the screen saver is inhibited, and
\fBidleDeinhibitCommand\fP is a command that is run once such a client
stops inhibiting the screen saver or loses the input focus.
.IP
Since there are several different screen saver inhibition protocols
used by different X desktops, the protocol translator refrains from
implementing any of them.  Instead, it provides the ability to run any
command necessary to inhibit and de-inhibit the screen saver.  For
example, an
.BR XScreenSaver(1)
user may wish to apply the following resources:
.IP
.in +4
.EX
\fB12to11.IdleInhibitCommand: xscreensaver-command -deactivate\fP
.EE
.in
.IP
The commands specified are expected to finish.  If the process
corresponding to one command is still running by the time another
command should have been run, the other command will be queued until
the first process exits.
.TP
.B idleCommandInterval\fP (class \fBIdleCommandInterval\fP)
The number of seconds the protocol translator waits between
invocations of the command specified in \fBidleIntervalCommand\fP.
Defaults to 60.
.TP
.B additionalModifiersOfScreen\fP\fINUMBER\fP (class \fBAdditionalModifiers\fP)
A comma-separated list of additional DRM modifiers that can be applied
to each format.  These modifiers must be specified by their names
inside the `drm_fourcc.h' system header, and should be used if your X
graphics driver fails to report which modifiers are available.
\fINUMBER\fP should be replaced with the number of the screen on which
the protocol translator is running.

On 9th Generation Intel systems, for example, you may want to set this
to:
.IP
.in +4
.EX
\fB12to11.AdditionalModifiers: I915_FORMAT_MOD_X_TILED,DRM_FORMAT_MOD_LINEAR\fP
.EE
.in
.IP
.SH ENVIRONMENT
Several environment variables exist that modify the behavior of the
protocol translator in one way or another.  Most of these are used for
debugging, but some may be relevant to users as well.
.PP
The
.B USE_BUILTIN_RESIZE
environment variable, if set, forces the use of the built-in
drag-to-resize support provided by the protocol translator, even if
the X window manager provides its own.
.PP
The
.B DEBUG_REFRESH_PREDICTION
environment variable, if set, forces the frame clock to predict the
presentation deadline of the X compositing manager.  This is used to
debug code that is otherwise not run without subsurfaces being
present.
.PP
The
.B DISABLE_FRAME_SYNCHRONIZATION
environment variable, if set, disables frame synchronization with the
X compositing manager.  Setting this variable is probably not a good
idea.
.PP
The
.B SYNCHRONIZE
environment variable, if set, causes the X library to check for errors
immediately after issuing a request.  The resulting backtraces from
the error handler then reflect the protocol request that actually
caused the error, instead of some unpredictable request in the future.
.PP
The
.B APPLY_STATE_WORKAROUND
environment variable, if set, causes the protocol translator to handle
toplevel window state changes differently.  If Wayland programs do not
make the transition between fullscreen and maximized states correctly,
try setting this variable.
.PP
The
.B GLOBAL_SCALE
environment variable, if set to an integer, overrides the global
program scale factor normally provided by the
.I Gdk/WindowScalingFactor
X setting.
.PP
The
.B OUTPUT_SCALE
environment variable, if set to an integer, overrides the output scale
factor that is otherwise set to the global program scale factor.
Setting this option is only useful to debug Wayland program
downscaling.
.PP
The
.B DIRECT_STATE_CHANGES
variable, if set, forces ConfigureNotify events from the window
manager to be handled directly, without waiting some time for a
corresponding _NET_WM_STATE event to arrive.  This is only
useful for debugging the window resizing logic.
.PP
The
.B RENDERER
variable, if set, controls the rendering backend used by the
protocol translator.  When set to
.I help
it prints a list of available rendering backends instead.  This takes
precedence over the \fBrenderer\fP resource whenever set.
.PP
The
.B RENDER_VISUAL
variable, if set to a number, contains the ID of the visual used
when the EGL rendering backend is in use.
.SH "CONFORMING TO"
The protocol translator aims to comply with the specifications of the
following Wayland interfaces:
.TS H
lb lb
lb n .
Protocol	Version
wl_output	4
wl_compositor	5
wl_shm	1
xdg_wm_base	5
wl_subcompositor	1
wl_seat	8
wl_data_device_manager	3
zwp_linux_dmabuf_v1	4
zwp_primary_selection_device_manager_v1	1
wp_viewporter	1
zxdg_decoration_manager_v1	1
zwp_text_input_manager_v3	1
wp_single_pixel_buffer_manager_v1	1
zwp_pointer_constraints_v1	1
zwp_relative_pointer_manager	1
zwp_idle_inhibit_manager_v1	1
xdg_activation_v1	1
wp_tearing_control_manager_v1	1
.TE
.PP
When the protocol translator is built with EGL support, the following
protocol is also supported:
.TS H
lb lb
lb n .
Protocol	Version
zwp_linux_explicit_synchronization_v1	2
.TE
.PP
When the X server supports version 1.6 or later of the X Resize,
Rotate and Reflect Extension, the following Wayland protocol is also
supported:
.TS H
lb lb
lb n .
Protocol	Version
wp_drm_lease_device_v1	1
.TE
.PP
When the X server supports version 2.4 or later of the X Input
Extension, the following Wayland protocol is also supported:
.TS H
lb lb
lb n .
Protocol	Version
zwp_pointer_gestures_v1	3
.TE
.PP
However, Wayland clients are allowed to continue to access data from
the \fBCLIPBOARD\fP and \fBPRIMARY\fP selections even when they do not
have the keyboard focus, against the restrictions put out in the
relevant protocol specifications.  It is the opinion of the authors
that such ``security'' does not come with any real benefit.
.SH BUGS
There is a hard to catch bug where Wayland programs leaving the
fullscreen or maximized state may abruptly return to their maximized
size.  Setting the
.B APPLY_STATE_WORKAROUND
environment variable may help.
.PP
Using this protocol translator under a window manager that does not at
least support the
.B _NET_WM_SYNC_REQUEST
window manager protocol will result in Wayland programs running badly.
.PP
.SH "SEE ALSO"
X(7), Xorg(1)
.SH AUTHORS
Various contributors.


12to11.c
/* Wayland compositor running on top of an X serer.

Copyright (C) 2022 to various contributors.

This file is part of 12to11.

12to11 is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

12to11 is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with 12to11.  If not, see <https://www.gnu.org/licenses/>.  */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <time.h>
#include <locale.h>

#include "compositor.h"


/* Globals.  */
Compositor compositor;

static void
DetermineServerTime (void)
{
  Time server_time;
  struct timespec clock_spec;
  uint64_t clock_ms, truncated;

  /* Try to determine if the X server time is the same as the
     monotonic time.  If it is not, certain features such as "active"
     frame synchronization will not be available.  */

  clock_gettime (CLOCK_MONOTONIC, &clock_spec);
  server_time = XLGetServerTimeRoundtrip ();

  /* Convert the monotonic time to milliseconds.  */

  if (IntMultiplyWrapv (clock_spec.tv_sec, 1000, &clock_ms))
    goto overflow;

  if (IntAddWrapv (clock_ms, clock_spec.tv_nsec / 1000000,
		   &clock_ms))
    goto overflow;

  /* Truncate the time to 32 bits.  */
  truncated = clock_ms;

  /* Compare the clock time with the server time.  */
  if (llabs ((long long) server_time - (long long) truncated) <= 5)
    /* Since the difference between the server time and the monotonic
       time is less than 5 ms, the server time is most likely the
       monotonic time.  */
    compositor.server_time_monotonic = True;
  else
    {
    overflow:
      compositor.server_time_monotonic = False;
      fprintf (stderr, "Warning: the X server time does not seem to"
	       " be synchronized with the monotonic time.  Multiple"
	       " subsurfaces may be displayed at a reduced maximum"
	       " frame rate.\n");
    }
}

static void
HandleCmdline (Display *dpy, const char *socket, int argc, char **argv)
{
  int i;
  XrmDatabase rdb, initial_rdb;

  /* Set the default resource and class names.  */
  compositor.resource_name = "12to11";
  compositor.app_name = "12to11";

  if (argc < 2)
    /* There are no arguments to handle.  */
    return;

  /* Obtain the resource database.  */
  initial_rdb = rdb = XrmGetDatabase (dpy);

  /* Determine the instance name based on the executable.  First,
     remove any leading directory separator from argv[0].  */
  compositor.app_name = strrchr (argv[0], '/');

  /* If no directory separator was present, just use it.  */
  if (!compositor.app_name)
    compositor.app_name = argv[0];
  else
    /* Otherwise, strip the trailing '/'.  */
    compositor.app_name = compositor.app_name + 1;

  for (i = 1; i < argc; ++i)
    {
      if (!strcmp (argv[i], "-help"))
	{
	print_usage:
	  fprintf (stderr,
		   "usage: %s [-name name] [-class class] [-printsocket]"
		   " [-xrm resourcestring...]\n",
		   argv[0]);
	  exit (!strcmp (argv[i], "-help") ? 0 : 1);
	}
      else if (!strcmp (argv[i], "-class"))
	{
	  if (i + 1 >= argc)
	    {
	      fprintf (stderr, "%s: option -class requires a value\n",
		       argv[0]);
	      exit (1);
	    }

	  compositor.resource_name = argv[++i];
	}
      else if (!strcmp (argv[i], "-name"))
	{
	  if (i + 1 >= argc)
	    {
	      fprintf (stderr, "%s: option -name requires a value\n",
		       argv[0]);
	      exit (1);
	    }

	  compositor.app_name = argv[++i];
	}
      else if (!strcmp (argv[i], "-xrm"))
	{
	  if (i + 1 >= argc)
	    {
	      fprintf (stderr, "%s: option -xrm requires a value\n",
		       argv[0]);
	      exit (1);
	    }

	  XrmPutLineResource (&rdb, argv[++i]);
	}
      else if (!strcmp (argv[i], "-printsocket"))
	puts (socket);
      else
	{
	  fprintf (stderr, "%s: bad command line option \"%s\"\n",
		   argv[0], argv[1]);
	  goto print_usage;
	}
    }

  /* In case XrmPutLineResource created a new database, set it as the
     display's resource database.  */
  if (rdb != initial_rdb)
    XrmSetDatabase (dpy, rdb);
}

static void
XLMain (int argc, char **argv)
{
  Display *dpy;
  struct wl_display *wl_display;
  const char *socket;

  /* Set the locale.  */
  setlocale (LC_ALL, "");

  /* Initialize Xlib threads.  */
  XInitThreads ();

  dpy = XOpenDisplay (NULL);
  wl_display = wl_display_create ();

  if (!dpy || !wl_display)
    {
      fprintf (stderr, "Display initialization failed\n");
      exit (1);
    }

  socket = wl_display_add_socket_auto (wl_display);

  if (!socket)
    {
      fprintf (stderr, "Unable to add socket to Wayland display\n");
      exit (1);
    }

  /* Call XGetDefault with some dummy values to have the resource
     database set up.  */
  XrmInitialize ();
  XGetDefault (dpy, "dummmy", "value");

  /* Parse command-line arguments.  */
  HandleCmdline (dpy, socket, argc, argv);

  compositor.display = dpy;
  compositor.conn = XGetXCBConnection (dpy);
  compositor.wl_display = wl_display;
  compositor.wl_socket = socket;
  compositor.wl_event_loop
    = wl_display_get_event_loop (wl_display);

  /* Initialize server time tracking very early.  */
  InitTime ();

  InitXErrors ();
  SubcompositorInit ();
  InitSelections ();

  XLInitTimers ();
  XLInitAtoms ();

  /* Initialize renderers immediately after timers and atoms are set
     up.  */
  InitRenderers ();

  XLInitRROutputs ();
  XLInitCompositor ();
  XLInitSurfaces ();
  XLInitShm ();
  XLInitXdgWM ();
  XLInitXdgSurfaces ();
  XLInitXdgToplevels ();
  XLInitFrameClock ();
  XLInitSubsurfaces ();
  XLInitSeats ();
  XLInitDataDevice ();
  XLInitPopups ();
  XLInitDmabuf ();
  XLInitXData ();
  XLInitXSettings ();
  XLInitIconSurfaces ();
  XLInitPrimarySelection ();
  XLInitExplicitSynchronization ();
  XLInitWpViewporter ();
  XLInitDecoration ();
  XLInitTextInput ();
  XLInitSinglePixelBuffer ();
  XLInitDrmLease ();
  XLInitPointerConstraints ();
  XLInitRelativePointer ();
  XLInitKeyboardShortcutsInhibit ();
  XLInitIdleInhibit ();
  XLInitPointerGestures ();
  XLInitXdgActivation ();
  XLInitTearingControl ();
  XLInitTest ();

  /* This has to come after the rest of the initialization.  */
  DetermineServerTime ();
  XLRunCompositor ();
}

int
main (int argc, char **argv)
{
  XLMain (argc, argv);
  return 0;
}
